riscv-semihosting/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+//! [examples/semihosting.rs:40] i = 0
+//! [examples/semihosting.rs:41] UUID.offset(i as isize).read_volatile() = 3370045464
+//! [examples/semihosting.rs:40] i = 1
+//! [examples/semihosting.rs:41] UUID.offset(i as isize).read_volatile() = 1426218275
+//! [examples/semihosting.rs:40] i = 2
+//! [examples/semihosting.rs:41] UUID.offset(i as isize).read_volatile() = 2422621116
+//! [examples/semihosting.rs:40] i = 3
+//! [examples/semihosting.rs:41] UUID.offset(i as isize).read_volatile() = 1044138593
+//! ```
+//!
+//! # Optional features
+//!
+//! ## `jlink-quirks`
+//!
+//! When this feature is enabled, return values above `0xfffffff0` from
+//! semihosting operation `SYS_WRITE` (0x05) are interpreted as if the entire
+//! buffer had been written. The current latest version 6.48b of J-Link exhibits
+//! such behaviour, causing a panic if this feature is not enabled.
+//!
+//! ## `no-semihosting`
+//!
+//! When this feature is enabled, the underlying system calls are patched out.
+//!
+//! # Reference
+//!
+//! For documentation about the semihosting operations, check
+//! ['Semihosting for AArch32 and AArch64'](https://github.com/ARM-software/abi-aa/blob/main/semihosting/semihosting.rst).
+//! The RISC-V Semihosting spec is identical to Arm's with the exception of the
+//! assembly sequence necessary to trigger a semihosting call, so their
+//! documentation is sufficient.
+
+#![deny(missing_docs)]
+#![no_std]
+
+#[cfg(all(riscv, not(feature = "no-semihosting")))]
+use core::arch::asm;
+
+#[macro_use]
+mod macros;
+
+pub mod debug;
+#[doc(hidden)]
+pub mod export;
+pub mod hio;
+pub mod nr;
+
+/// Performs a semihosting operation, takes a pointer to an argument block
+///
+/// # Safety
+///
+/// The syscall number must be a valid [semihosting operation],
+/// and the arguments must be valid for the associated operation.
+///
+/// [semihosting operation]: https://developer.arm.com/documentation/dui0471/i/semihosting/semihosting-operations?lang=en
+#[inline(always)]
+pub unsafe fn syscall<T>(nr: usize, arg: &T) -> usize {
+    syscall1(nr, arg as *const T as usize)
+}
+
+/// Performs a semihosting operation, takes one integer as an argument
+///
+/// # Safety
+///
+/// Same as [`syscall`].
+#[inline(always)]
+pub unsafe fn syscall1(_nr: usize, _arg: usize) -> usize {
+    match () {
+        #[cfg(all(riscv, not(feature = "no-semihosting")))]
+        () => {
+            let mut nr = _nr;
+            // The instructions below must always be uncompressed, otherwise
+            // it will be treated as a regular break, hence the norvc option.
+            //
+            // See https://github.com/riscv/riscv-semihosting-spec for more details.
+            asm!("
+                .option push
+                .option norvc
+                slli x0, x0, 0x1f
+                ebreak
+                srai x0, x0, 0x7
+                .option pop
+            ",
+            inout("a0") nr,
+            in("a1") _arg,
```

Comment:
There appear to be two bugs in this asm call.

First, [riscv-semihosting-spec](https://github.com/riscv-software-src/riscv-semihosting/blob/f2e0eaf46d34a54ebbc1799c9a8879c4cc544fe4/riscv-semihosting-spec.adoc#21-semihosting-trap-instruction-sequence) says:

> If paging is in use in the current mode, this sequence must not cross a page boundary as the semihosting system must be able to check for the semihosting sequence without needing data from potentially missing pages.

I believe you have to use `.balign 16`.

Second, some syscalls such as [`SYS_ELAPSED`](https://github.com/ARM-software/abi-aa/blob/HEAD/semihosting/semihosting.rst#633return) change `a1`, so I believe you have to use `inout` instead of `in` for `a1`.

See also my [semihosting crate](https://github.com/taiki-e/semihosting/blob/969d2350d51c38740649837f09e23a5a0de64cb8/src/sys/arm_compat/syscall/riscv.rs#L12-L25) for an example of RISC-V semihosting syscalls in Rust.